### ISSUE TITLE
When I Inject forza horizon 4 the game crash？

### ISSUE THREAD

**Issue by LRW-Adolph** [araghon007/X1nput#30]
When I Inject forza horizon 4 the game crash？

**Comment by SojerManLan** [comment]
> When I Inject forza horizon 4 the game crash？

Forza Horizon 4, like many Microsoft Store games, makes use of Arxan Anti-Tamper DRM, which will not allow any injections (namely, games using it will crash if an injection is detected). Besides, Forza Horizon 4 already includes support for Impulse Triggers, so using X1nput with the game is not needed.

**Comment by araghon007** [comment]
> Forza Horizon 4, like many Microsoft Store games, makes use of Arxan Anti-Tamper DRM, which will not allow any injections (namely, games using it will crash if an injection is detected).

Oh, thanks. I remember also trying it because of some other issue someone reported and I was wondering why it was crashing. I assumed it had some anti-tamper stuff, but wasn't sure what exactly.

I think I managed to get it working by just copying the X1nput DLL into the game's directory and renaming it to one of the XAudio versions, but I can't remember which one. Has to be XAudio, since only one XInput DLL is loaded, and the way this code works, it requires a real XInput DLL to be loaded. Just check which DLLs get loaded by the game, and try renaming the X1nput DLL to one of them.

I was thinking to doing this automatically as a way to allow automatically running it when you start the game, but I'm not sure what the best way would be. I could either create a sort of redirect thing, where XInput calls get redirected to a real XInput DLL and also keep the detour stuff, since that would be the easiest, or rewrite the XInput DLL from scratch, which would be way harder as I wasn't able to find any code for it, plus I don't think I'd be able to send a HID report as easily as I can now, because handles and stuff.

**Comment by Squall-Leonhart** [comment]
FH4 natively supports impulse triggers, it doesn't require this tool.

**Comment by araghon007** [comment]
It only natively supports them only on the UWP version, doesn't it?

**Comment by SojerManLan** [comment]
Yes, it does support them natively on the Microsoft Store version.

**Comment by araghon007** [comment]
Right, but not the Steam version, which is what I assumed was talked about here, since to my knowledge it's pretty hard to even try to inject anything into UWP apps

**Comment by Squall-Leonhart** [comment]
it natively uses them on both versions.

**Comment by araghon007** [comment]
Are you 100% sure? From what I've tested about a year ago, they didn't work.

**Comment by YuanzheGu** [comment]
Still doesn't support impulse trigger on steam version, and apparently they're not gonna fix it. Add this in fh5 however.
